### DISCUSSION TITLE
How to do a general magic link / redirect if magic link is expired

### DISCUSSION THREAD

**Discussion by Brentably** [supabase/supabase#30660] [Questions]
I'm using NextJS pages. 

I have a nice magic link login flow working fine. However, I want to add some routing to my application. If the user is logged in, and they're on the home page, I want to take them to the "/dashboard" page. If the user is on the dashboard page, but they're not logged in, I want to take them to "/" so they can log in. 

This seems like some pretty straightforward functionality, but I can't get it working. 

My best idea is to use a middleware to check if a user is logged in.

Great so I'm using the code I found here:

[https://github.com/supabase/supabase/blob/d2facb942c0740d985908468d3817f951463e40b/examples/user-management/nextjs-user-management/utils/supabase/middleware.ts](https://github.com/supabase/supabase/blob/d2facb942c0740d985908468d3817f951463e40b/examples/user-management/nextjs-user-management/utils/supabase/middleware.ts) 

to authenticate in the middleware and see if a user is logged in or not, so I can route them where I need to go.

Here is my middleware.ts code: 
```
import { createServerClient } from "@supabase/ssr";
import { NextResponse, type NextRequest } from "next/server";

export async function middleware(request: NextRequest) {
  let supabaseResponse = NextResponse.next({
    request,
  });

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return request.cookies.getAll();
        },
        setAll(cookiesToSet) {
          cookiesToSet.forEach(({ name, value, options }) =>
            request.cookies.set(name, value)
          );
          supabaseResponse = NextResponse.next({
            request,
          });
          cookiesToSet.forEach(({ name, value, options }) =>
            supabaseResponse.cookies.set(name, value, options)
          );
        },
      },
    }
  );

  // refreshing the auth token
  const { data: { user } } = await supabase.auth.getUser();
  console.log('user', user);

  const shouldRedirectToPractice = user && request.nextUrl.pathname === "/";
  if(shouldRedirectToPractice) {
    return NextResponse.redirect(new URL("/practice", request.url));
  }
  const shouldRedirectToHome = !user && request.nextUrl.pathname !== "/";
  if(shouldRedirectToHome) {
    console.log('redirecting to home');
    console.log('request.nextUrl.pathname', request.nextUrl.pathname);
    return NextResponse.redirect(new URL("/", request.url));
  }

  return supabaseResponse;
}


export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * Feel free to modify this pattern to include more paths.
     */
    "/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)",
  ],
};
```

Unfortunately this interrupts the validation / login flow from magic link, because it detects that there is no user. 
I can hack around this by putting:
```
  if(request.nextUrl.toString().includes("/practice?code")) {
    return NextResponse.next();
  }
```

Near the top of the document, but, ideally I can handle the edge case where the user clicks a dead magic link, and redirect them to a page that tells them the magic link is dead. How would I go about doing this?

**Comment by silentworks** [comment]
Why not move your dead magic link page to a different page? one that doesn't require authentication.
